### ISSUE TITLE
Tooltips not appearing on OS X (10.12.2)

### ISSUE THREAD

**Issue by theyowguy** [pyfa-org/Pyfa#909]
Hovering over icons in pyfa does not show any tooltips on OS X.

**Comment by Ebag333** [comment]
>got it running from source… no tooltips

Seems like it doesn't work from the skeleton or source.

**Comment by blitzmann** [comment]
pyfa kinda sorta doesn't work well with 10.12. Are all tooltips broken, or just specific ones?

Unfortunately I have no way of fixing this. working with OS X is a bitch and a half, and we don't have a developer that's specifically focused on OS X.

**Comment by theyowguy** [comment]
I'm not seeing any tooltips at all.

Running either the binary or from source having installed the dependencies using homebrew and pip.

I will try to find a windows machine to use so I can see the expected behaviour.

**Comment by theyowguy** [comment]
I'm new to the game and pyfa, I was just looking for an easy way to figure out what all the icons mean. I'm told they match those in the game so I can use that to learn them.

**Comment by blitzmann** [comment]
yeah, unfortunately like I said, we don't have a OS X dev. And the GUI toolkit we use (wxPython) has some issues with OS X and all it's different versions, 10.11+ being the worse.

> from source having installed the dependencies using homebrew and pip.

it sounds like you're a dev? hint hint ;)

**Comment by theyowguy** [comment]
It has been many years since I've played around with Python and never with a GUI toolkit. I see what you mean about wxPython having issues on OS X... I wasn't able to get versions from the wxpython.org page installed on 10.12.

I managed to get pyfa running on windows and I see everything working as it should which will be a help.

Let me play around for a bit and hopefully I can come up with something.

Cheers.

**Comment by blitzmann** [comment]
http://davixx.fr/blog/2016/01/25/wxpython-on-os-x-el-capitan/

May be of use at some point
The latest VM I have of OS X is 10.10. I was never able to find a good hackintosh image for 10.11 or 10.12, so I can't test on those. =(

**Comment by theyowguy** [comment]
It appears homebrew does have the same version;

wxpython: stable 3.0.2.0 (bottled)
Python bindings for wxWidgets
https://www.wxwidgets.org/
/usr/local/Cellar/wxpython/3.0.2.0 (1,107 files, 38.1M) *
  Poured from bottle on 2016-12-16 at 19:03:09
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wxpython.rb
==> Dependencies
Required: wxmac ✔
==> Options
--universal
	Build a universal binary

Do you think the version of python would / will matter? I'm using;

Python 2.7.12

**Comment by blitzmann** [comment]
Latest 2.7 is recommended. 3.0+ doesn't work. You should be good :)

**Comment by Ebag333** [comment]
>Latest 2.7 is recommended. 3.0+ doesn't work. You should be good :)

We aren't very far off from making 3.0 work according to @a-tal . But that's not going to fix the tooltip issue. :(

Does the 2.9 version of wxPython not run at all on OSx 10.12?

**Comment by theyowguy** [comment]
I don't think the problem is with WxPython. My guess is somewhere along the lines OS X changed the way tooltips work and those changes haven't been picked up by WxPython (although I may be completely out to lunch as well!!!).

The WxPython code is from Sept. 2014 but I did find a beta of the package from Sept. 2016 which I couldn't get to run - the app would fail on import wx.

**Comment by blitzmann** [comment]
@ebag333 Even if pyfa technically could be upgraded to py3, that's not something we are doing for a variety of reasons.

As for wxpython , it may be that wxpython (rather, wxWidgets) hasn't kept up, in which  are there's nothing we can do about it. :(

**Comment by Ebag333** [comment]
>but I did find a beta of the package from Sept. 2016 which I couldn't get to run - the app would fail on import wx

That might be wx Phoenix. I played around briefly with it, but it is massively different and would take a lot of effort to convert over to it. 

They do have a list conversions, but it's not complete. And there is stuff we do that they don't handle at all yet.

**Comment by theyowguy** [comment]
> That might be wx Phoenix. I played around briefly with it, but it is massively different and would take a lot of effort to convert over to it.

Don't think so... it was version 3.0.3.0 (wxPython3.0-osx-3.0.3.0.b20160914-cocoa-py2.7)

**Comment by Ebag333** [comment]
I think that is wx Phoenix.  Some of their documentation mentions 3.0.3. 

https://wxpython.org/Phoenix/docs/html/main.html

**Comment by Ebag333** [comment]
wxPython is essentially dead, Phoenix has taken it over. So anything recently released is going to be Phoenix. 

here is the conversion guide. 
https://wxpython.org/Phoenix/docs/html/MigrationGuide.html

**Comment by resinneublem** [comment]
I'll try and help take a look too. I still have access to my mac. (I'm currently moving to ubuntu)

For what it's worth, I just now followed the directions on http://www.wikigain.com/install-macos-sierra-10-12-virtualbox/ and got macOS Sierra working in VirtualBox on Windows without a problem. In case anyone needs to get a local development VM up and running.

**Comment by resinneublem** [comment]
So, I sorta wonder if tooltips are just straight up not implemented in macOS.

I'm seeing some stuff here and there saying they don't work, but this is chatter from years ago.

> Regular tooltips do work on wxMac...

https://groups.google.com/forum/#!msg/wxpython-mac/79z0ovbCUnA/C4ZWL_3EAboJ from 2006

> wxtooltip is not implemented in osx_cocoa yet

http://osdir.com/ml/wx-users/2010-01/msg00416.html 2010

---

I found https://wiki.wxpython.org/AnotherTutorial#wx.Window and just re-ran their demo Ubuntu vs macOS and the tooltips don't show up.

```python
#!/usr/bin/python

# simple2.py

import wx

app = wx.App()

frame = wx.Frame(None, -1, '')
frame.SetToolTip(wx.ToolTip('This is a frame'))
frame.SetCursor(wx.StockCursor(wx.CURSOR_MAGNIFIER))
frame.SetPosition(wx.Point(0,0))
frame.SetSize(wx.Size(300,250))
frame.SetTitle('simple2.py')
frame.Show()

app.MainLoop()
```

What's weird is that when the app opens, I sometimes see a tooltip in macOS in the main fitting area when there's not a fit selected.

**Comment by theyowguy** [comment]
Interesting... I never considered they might not have ever worked. I had just assumed that they had been working and they changed something in macOS or a previous version which "broke" things.

**Comment by blitzmann** [comment]
I wonder if this is something that happened with 10.12 or an earlier version... I know 10.11 caused a lot of issues with wxWidgets.

I also wonder if it's possible to set out own tooltip class for OS X (or maybe for all platforms), where we override the default tooltip behavior and go with one that's implemented in python. IIRC there are a few tooltip classes that come with wxPython
